src/nsrt_learning/side_predicate_learning.py

```diff
@@ -484,6 +455,61 @@ def _sideline(self) -> List[PartialNSRTAndDatastore]:
         assert self._check_harmlessness(tuple(param_opt_to_pnad.values()))
         return list(param_opt_to_pnad.values())
 
+    def _try_specifizing_pnad(
+        self,
+        necessary_add_effects: Set[GroundAtom],
+        pnad: PartialNSRTAndDatastore,
+        segment: Segment,
+    ) -> Optional[PartialNSRTAndDatastore]:
+        """Given a PNAD and some necessary add effects that the PNAD must
+        achieve, try to make the PNAD's add effects more specific ("specifize")
+        so that they cover these necessary add effects.
+
+        Returns the new constructed PNAD, without modifying the
+        original. If the PNAD does not have a grounding that can even
+        partially satisfy the necessary add effects, returns None.
+        """
+
+        # Get an arbitrary grounding of the PNAD's operator whose
+        # preconditions hold in segment.init_atoms and whose add
+        # effects are a subset of necessary_add_effects.
+        ground_op = self._get_partially_satisfying_grounding(
+            necessary_add_effects, pnad, segment)
+        # If no such grounding exists, specifizing is not possible.
+        if ground_op is None:
+            return None
+        # To figure out the effects we need to add to this PNAD,
+        # we first look at the ground effects that are missing
+        # from this arbitrary ground operator.
+        missing_effects = (necessary_add_effects - ground_op.add_effects)
+        obj_to_var = dict(zip(ground_op.objects, pnad.op.parameters))
+        # Before we can lift missing_effects, we need to add new
+        # entries to obj_to_var to account for the situation where
+        # missing_effects contains objects that were not in
+        # the ground operator's parameters.
+        var_names = [var.name for var in obj_to_var.values()]
+        assert all(name.startswith("?x") for name in var_names)
+        max_var_name = max(var_names,
+                           key=lambda name: int(name[2:]),
+                           default="?x-1")
+        max_var_num = int(max_var_name[2:])
+        for effect in missing_effects:
+            for obj in effect.objects:
+                if obj not in obj_to_var:
+                    max_var_num += 1
+                    new_var = Variable("?x" + str(max_var_num), obj.type)
```

Comment:
sorry I missed this before. so now it looks like we have dashes in between ?x and the number. should we add a dash here too then?

Reply:
no, the `?x-1` is a negative one. so that when we increment, we start at 0 as intended. i'll add a comment

Reply:
oh i see. okay, I definitely want to do the utility function refactor that I mentioned then!